cdap-app-fabric/src/main/java/co/cask/cdap/app/stream/DefaultStreamWriter.java

```diff
@@ -80,6 +97,15 @@ private URL getStreamURL(String stream, boolean batch) throws IOException {
   }
 
   private void writeToStream(Id.Stream stream, HttpRequest request) throws IOException {
+    if (!isStreamRegistered.containsKey(stream)) {
```

Comment:
So there is race here if there are concurrent calls to `writeToStream` (both threads `containsKey` return false, they'll both register and both update the map). I guess it's ok to make more than one register call, but strictly speaking it's promoting a wrong assumption here. If you want to make it atomic, you should use `putIfAbsent`